WinCompatibilityPack/WinCompatibilityPack.psm1

```diff
@@ -0,0 +1,804 @@
+                -ConfigurationName $configurationName @SessionOpts
+    }
+    else
+    {
+        Write-Verbose -Verbose:$verboseFlag "Reusing the existing compatibility session."
+    }
+
+    if ($PassThru)
+    {
+        return $session
+    }
+}
+
+###########################################################################################
+<#
+.Synopsis
+   This command defines a global function that always runs in the compatibility session.
+.DESCRIPTION
+    This command defines a global function that always runs in the compatibility session,
+    returning serialized data to the calling session. Parameters can be specified using
+    the 'param' statement but only positional parameters are supported.
+
+    By default, when executing, the current compatibility session is used,
+    or, in the case where there is no existing session, a new default session
+    will be created. This behaviour can be overridden using the
+    additional parameters on the command.
+.EXAMPLE
+    Add-WinFunction myFunction {param ($n) "Hi $n!"; $PSVersionTable.PSEdition }
+    This example defines a function called 'myFunction' with 1 parameter. When invoked it will print a message then return the PSVersion table
+    from the compatibility session. Now call the function
+    PS C:\> myFunction Bill
+    Hi Bill!
+    Desktop
+#>
+function Add-WinFunction
+{
+    [CmdletBinding()]
+    [Alias("awinf")]
+    [OutputType([void])]
+    Param
+    (
+        # The name of the function to define
+        [Parameter(Mandatory,Position=0)]
+        [String]
+            $FunctionName,
+
+        # Scriptblock to use as the body of the function
+        [Parameter(Mandatory,Position=1)]
+        [ScriptBlock]
+            $ScriptBlock,
+
+        # If you don't want to use the default compatibility session, use
+        # this parameter to specify the name of the computer on which to create
+        # the compatibility session.
+        # (Defaults to 'localhost')
+        [Parameter()]
+        [String]
+        [Alias("Cn")]
+            $ComputerName = $script:DefaultComputerName,
+
+        # Specifies the configuration to connect to when creating the compatibility session
+        # (Defaults to 'Microsoft.PowerShell')
+        [Parameter()]
+        [String]
+            $ConfigurationName = $script:DefaultConfigurationName,
+
+        # The credential to use when creating the compatibility session
+        # using the target machine/configuration
+        [Parameter()]
+        [PSCredential]
+            $Credential
+    )
+
+    # Make sure the session is initialized
+    [void] $PSBoundParameters.Remove('FunctionName')
+    [void] $PSBoundParameters.Remove('ScriptBlock')
+
+    Initialize-WinSession @PSBoundParameters
+
+    $scriptToRun = @"
+    param()
+    
+    `$session = Get-PSsession -Name ${script:SessionName}
+    Invoke-Command -Session `$session -Scriptblock `$ScriptBlock -ArgumentList `$args
+"@
```

Comment:
This here string requires more escapes than string interpolation. I recommend switching to `@'` and using `-f` Also recommend enclosing the result of `$script:SessionName` in quotes.

Reply:
Since i'm eventually binding a closure anyway, there is no actual need for the string. I just made SessionName available in the local scope before calling CreateNewClosure() so it will be captured.